### PR TITLE
[RFR] Refactoring and adding new update testing

### PR DESCRIPTION
**Pull request by lcouzens**
Cleaning up current tests and expanding on upgrade/update coverage.

The current errors all seem to be issues with template provisions and ssh issues.

### CODE REVIEW

**Comment by dajoRH** [comment]
Lint report for commit a52c705d03872073ea6b5ec498a377dd26d64c12:

`cfme/tests/cli/test_appliance_update.py`:
- :heavy_exclamation_mark: Line 534:1: **W391** *blank line at end of file*

`cfme/tests/cli/test_appliance_upgrade.py`:
- :red_circle: Line 15:1: **F401** *'cfme.utils.version.get_stream' imported but unused*
- :heavy_exclamation_mark: Line 188:57: **W292** *no newline at end of file*

`cfme/tests/test_db_migrate.py`:
- :red_circle: Line 1:1: **F401** *'tempfile' imported but unused*

Please, rectify these issues :smirk: .


**No commit flaws detected**.

*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit d839b323982c2a09dccff456b07ac94486bcccaf:

`cfme/tests/cli/test_appliance_update.py`:
- :red_circle: Line 21:1: **F401** *'cfme.utils.update.update' imported but unused*

`cfme/tests/cli/test_appliance_upgrade.py`:
- :red_circle: Line 19:1: **F401** *'cfme.utils.update.update' imported but unused*

Please, rectify these issues :smirk: .


**No commit flaws detected**.

*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit 83db91075faa7c085bf3d12910295a531f68cbda:

`cfme/tests/cli/test_appliance_update.py`:
- :red_circle: Line 10:1: **F401** *'cfme.fixtures.pytest_store.store' imported but unused*

Please, rectify these issues :smirk: .


**No commit flaws detected**.

*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit 97a0827fb379236ed8d43db9e807fba8c942a0d8:

`cfme/tests/cli/test_appliance_upgrade.py`:
- :bangbang: Line 169:5: **E303** *too many blank lines (3)*
- :bangbang: Line 190:101: **E501** *line too long (102 > 100 characters)*
- :bangbang: Line 580:35: **E202** *whitespace before ']'*

Please, rectify these issues :smirk: .


**No commit flaws detected**.

*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit 3ab4ecf409f21e70c4a5d7ff45f579987f75ee26:

`cfme/tests/cli/test_appliance_upgrade.py`:
- :heavy_exclamation_mark: Line 168:1: **W293** *blank line contains whitespace*

Please, rectify these issues :smirk: .


**No commit flaws detected**.

*CFME QE Bot*

File: cfme/tests/cli/test_appliance_update.py [hunk 1]
```diff
@@ -1,18 +1,28 @@
 import tempfile
-
+import fauxfactory
 import pytest
+import lxml.etree
+import yaml
```
Comment (quarckster) [inline cfme/tests/cli/test_appliance_update.py]: **REQUIRED**
Please sort imports in alphabetical order

File: cfme/tests/cli/test_appliance_update.py [hunk 2]
```diff
@@ -85,17 +95,236 @@ def appliance_preupdate(old_version, appliance):
     sp.destroy_pool(pool_id)
 
 
-@pytest.mark.uncollectif(lambda: not store.current_appliance.is_downstream)
-def test_update_yum(appliance_preupdate, appliance):
-    """Tests appliance update between versions"""
+@pytest.fixture(scope="function", )
```
Comment (quarckster) [inline cfme/tests/cli/test_appliance_update.py]: **REQUIRED**
Unneeded comma or you can even remove `(scope="function", )` at all.

File: cfme/tests/cli/test_appliance_update.py [hunk 3]
```diff
@@ -85,17 +95,236 @@ def appliance_preupdate(old_version, appliance):
...
+def multiple_preupdate_appliances(old_version, appliance):
+
+    series = appliance.version.series()
+    update_url = "update_url_{}".format(series.replace('.', ''))
+
+    """Requests appliance from sprout based on old_versions, edits partitions and adds
+    repo file for update. Returns two database-owning appliances, configures first appliance
+    with provider."""
```
Comment (quarckster) [inline cfme/tests/cli/test_appliance_update.py]: **REQUIRED**
This summary should be right after the function name.

File: cfme/tests/cli/test_appliance_update.py [hunk 4]
```diff
@@ -85,17 +95,236 @@ def appliance_preupdate(old_version, appliance):
...
+    command_set = ('ap', '', rep, '2', '1', '2', '', '', pwd, pwd, app0_ip, app1_ip, 'y',
+                   TimedCommand('y', 60), '')
+    apps[1].appliance_console.run_commands(command_set)
+    # Configure automatic failover on EVM appliance
+    command_set = ('ap', '', mon, TimedCommand('1', 30), '')
+    apps[2].appliance_console.run_commands(command_set)
+
+    def is_ha_monitor_started(appliance):
+        return bool(appliance.ssh_client.run_command(
+            "grep {} /var/www/miq/vmdb/config/failover_databases.yml".format(app1_ip)).success)
```
Comment (quarckster) [inline cfme/tests/cli/test_appliance_update.py]: **REQUIRED**
`success` returns a `bool` so no need to wrap bool again.

File: cfme/tests/cli/test_appliance_update.py [hunk 5]
```diff
@@ -127,20 +413,155 @@ def update_embedded_appliance(enabled_embedded_appliance, appliance):
...
+    """ Tests updating an appliance with providers, also confirms that the
+        provisioning continues to function correctly after the update has completed"""
+    update_appliance(appliance_with_providers)
+
+    wait_for(do_appliance_versions_match, func_args=(appliance, appliance_with_providers),
+             num_sec=900, delay=20, handle_exception=True)
+    # Verify that existing provider can detect new VMs on the second appliance
+    virtual_crud = provider_app_crud(VMwareProvider, appliance_with_providers)
+    vm = provision_vm(request, virtual_crud)
+    assert vm.provider.mgmt.is_vm_running(vm.name), "vm running"
```
Comment (quarckster) [inline cfme/tests/cli/test_appliance_update.py]: **REQUIRED**
A message after assertion should explain the error.
```python
assert vm.provider.mgmt.is_vm_running(vm.name), "vm is not running"
```

File: cfme/tests/cli/test_appliance_update.py [hunk 6]
```diff
@@ -127,20 +413,155 @@ def update_embedded_appliance(enabled_embedded_appliance, appliance):
...
             red_hat_updates.update_appliances()
-        return enabled_embedded_appliance
+        return appliance
+
+
+def do_appliance_versions_match(appliance1, appliance2):
+    """Checks if cfme-appliance has been updated"""
+    appliance2.rest_api._load_data()
+    try:
+        del appliance2.version
```
Comment (quarckster) [inline cfme/tests/cli/test_appliance_update.py]: **REQUIRED**
Please add a comment that you are clearing the cached value here.

**Comment by dajoRH** [comment]
Would you mind rebasing this Pull Request against latest master, please? :trollface:
*CFME QE Bot*

File: cfme/fixtures/cli.py [hunk 7]
```diff
@@ -118,3 +129,331 @@ def app_creds_modscope():
...
+    sp.destroy_pool(pool_id)
+
+
+@pytest.fixture
+def multiple_preupdate_appliances(old_version, appliance):
+
+    series = appliance.version.series()
+    update_url = "update_url_{}".format(series.replace('.', ''))
+
+    """Requests appliance from sprout based on old_versions, edits partitions and adds
```
Comment (rlbabyuk) [inline cfme/fixtures/cli.py]: Required: move docs after function name

File: cfme/fixtures/cli.py [hunk 8]
```diff
@@ -118,3 +129,331 @@ def app_creds_modscope():
         'sshlogin': credentials['ssh']['username'],
         'sshpass': credentials['ssh']['password']
     }
+
+
+@pytest.fixture
+def appliance_preupdate(old_version, appliance):
+    """Requests appliance from sprout based on old_versions, edits partitions and adds
+    repo file for update"""
+    series = appliance.version.series()
```
Comment (rlbabyuk) [inline cfme/fixtures/cli.py]: Looks like we have duplication of line 138-153 in 3 fixtures. We can move it out to separate fixture like:
```@pytest.fixture
def get_app(appliance, old_version):
    def _get_app(app_number):
        series = appliance.version.series()
        update_url = "update_url_{}".format(series.replace('.', ''))
        usable = []
        sp = SproutClient.from_config()
        available_versions = set(sp.call_method('available_cfme_versions'))
        for a in available_versions:
            if a.startswith(old_version):
                usable.append(Version(a))
        usable_sorted = sorted(usable, key=lambda o: o.version)
        try:
            apps, pool_id = sp.provision_appliances(count=app_number, preconfigured=True,
                                                    provider_type='rhevm',
                                                    lease_time=180, version=str(usable_sorted[-1]))
        except Exception as e:
            logger.exception("Couldn't provision appliance with following error:{}".format(e))
            raise SproutException('No provision available')
        return apps, pool_id
    return _get_app

File: cfme/tests/cli/test_appliance_update.py [hunk 9]
```diff
@@ -98,49 +62,133 @@ def test_update_yum(appliance_preupdate, appliance):
...
-        if red_hat_updates.platform_updates_available():
-            red_hat_updates.update_appliances()
-        return enabled_embedded_appliance
+@pytest.mark.ignore_stream("upstream")
+def test_update_webui(appliance_with_providers, appliance, request, old_version):
+    """ Tests updating an appliance with providers, also confirms that the
+        provisioning continues to function correctly after the update has completed"""
+    update_appliance(appliance_with_providers)
+
+    wait_for(do_appliance_versions_match, func_args=(appliance, appliance_with_providers),
```
Comment (rlbabyuk) [inline cfme/tests/cli/test_appliance_update.py]: Required: add a message with info what we are waiting for
parameter message='' 
Please add to all wait_for

File: cfme/tests/cli/test_appliance_update.py [hunk 10]
```diff
@@ -98,49 +62,133 @@ def test_update_yum(appliance_preupdate, appliance):
...
+    name = "example_{}".format(fauxfactory.gen_alpha())
+    description = "edited_{}".format(fauxfactory.gen_alpha())
+    repository = repositories.create(
+        name,
+        REPOSITORIES[0],
+        description=description)
+    view = navigate_to(repository, "Details")
+    refresh = view.toolbar.refresh.click
+    wait_for(
+        lambda: view.entities.summary("Properties").get_text_of("Status") == "successful",
```
Comment (rlbabyuk) [inline cfme/tests/cli/test_appliance_update.py]: Required: view.entities.summary("Properties").get_text_of("Status").lower()

File: cfme/tests/cli/test_appliance_update.py [hunk 11]
```diff
@@ -98,49 +62,133 @@ def test_update_yum(appliance_preupdate, appliance):
...
+    wait_for(do_appliance_versions_match, func_args=(appliance, ext_appliances_with_providers[0]),
+             num_sec=900, delay=20, handle_exception=True)
+    wait_for(do_appliance_versions_match, func_args=(appliance, ext_appliances_with_providers[1]),
+             num_sec=900, delay=20, handle_exception=True)
+    # Verify that existing provider can detect new VMs on both apps
+    virtual_crud_appl1 = provider_app_crud(VMwareProvider, ext_appliances_with_providers[0])
+    virtual_crud_appl2 = provider_app_crud(VMwareProvider, ext_appliances_with_providers[1])
+    vm1 = provision_vm(request, virtual_crud_appl1)
+    vm2 = provision_vm(request, virtual_crud_appl2)
+    assert vm1.provider.mgmt.is_vm_running(vm1.name), "vm is not running"
```
Comment (rlbabyuk) [inline cfme/tests/cli/test_appliance_update.py]: Required: use soft_assert fixture if you need to check vm2 is running, even if vm1 failed

File: cfme/tests/cli/test_appliance_upgrade.py [hunk 12]
```diff
@@ -0,0 +1,273 @@
...
+    name = "example_{}".format(fauxfactory.gen_alpha())
+    description = "edited_{}".format(fauxfactory.gen_alpha())
+    repository = repositories.create(
+        name,
+        REPOSITORIES[0],
+        description=description)
+    view = navigate_to(repository, "Details")
+    refresh = view.toolbar.refresh.click
+    wait_for(
+        lambda: view.entities.summary("Properties").get_text_of("Status") == "successful",
```
Comment (rlbabyuk) [inline cfme/tests/cli/test_appliance_upgrade.py]: Required: .lower(), correct all places

**Comment by dajoRH** [comment]
Lint report for commit 81442bdeded1fa9f8c559deedb2494530208f495:

`cfme/fixtures/cli.py`:
- :bangbang: Line 432:1: **E305** *expected 2 blank lines after class or function definition, found 0*
- :bangbang: Line 432:1: **E999** *SyntaxError: invalid syntax*
- :heavy_exclamation_mark: Line 432:2: **W292** *no newline at end of file*

Please, rectify these issues :smirk: .


**No commit flaws detected**.

*CFME QE Bot*

**Comment by dajoRH** [comment]
Would you mind rebasing this Pull Request against latest master, please? :trollface:
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit 6049cc5af36ee0c3c1165a0aa61025adac5238c1:

`cfme/fixtures/cli.py`:
- :heavy_exclamation_mark: Line 425:9: **W503** *line break before binary operator*

Please, rectify these issues :smirk: .


**No commit flaws detected**.

*CFME QE Bot*

**Comment by dajoRH** [comment]
Would you mind rebasing this Pull Request against latest master, please? :trollface:
*CFME QE Bot*

**Comment by izapolsk** [comment]
@lcouzens, could you please squash your PR. 41 commit is a bit too high for one PR.
In addition, it is difficult to review such PR. could you break it down into several smaller PRs ?

**Pull request closed without merging**